src/tools/miri/tests/pass/pattern-types.rs

```diff
@@ -0,0 +1,18 @@
+#![feature(pattern_types, pattern_type_macro, sized_hierarchy)]
+#![allow(dead_code)]
+
+use std::marker::PointeeSized;
+use std::mem::transmute;
+
+pub struct NonNull<T: PointeeSized> {
+    pointer: std::pat::pattern_type!(*const T is !null),
+}
+
+trait Trait {}
+impl Trait for () {}
+
+fn main() {
+    unsafe {
+        let _: NonNull<dyn Trait> = NonNull { pointer: transmute(&mut () as *mut dyn Trait) };
```

Comment:
@oli-obk the unsizing here happens before pattern types are even involved, doesn't it? It happens during the `&mut () as *mut dyn Trait`. So I don't see how this tests anything about pattern type unsizing.

Reply:
`NonNull<dyn Trait>` is a sized type, so I don't think there is an unsized pattern type in this example.